Implement per-instance authz

meatballhat opened the pull request:
Closes #19
- [x] copy token to separate key on initial fetch

meatballhat committed:
Implement per-instance authz

Closes #19

sarahhodne commented on server.go:
```diff
@@ -100,6 +106,38 @@ func (srv *server) requireAuth(w http.ResponseWriter, req *http.Request, next ht
+		return
+	}
+
+	if authHeader == fmt.Sprintf("token %s", instTok) {
```
Oooh, I like this for not accidentally falling through to the "permit" call :+1:

igorwwwwwwwwwwwwwwwwwwww replied:
Should use a constant time string comparison to prevent timing attacks.

meatballhat committed:
Back out func map lookup experiment

meatballhat committed:
Move token from temp location on access

and extend TTL on fetch

meatballhat committed:
Use constant time comparison for auth

meatballhat committed:
Comment out the SQS stuff until #5 can be addressed

meatballhat committed:
Add some tests for new db funcs

and a bonus test for the version printer

meatballhat committed:
Add some hacky content negotiation to token handler

so that tokens may be requested as text/plain to avoid the need for a json
parsing tool

meatballhat committed:
Break out separate TTL for instance tokens

as opposed to temporary instance tokens.  By doing this, we can keep the
vulnerable window open for a short time for initial instance token fetch, and
allow for a longer time window before receiving the first request authenticated
with the instance token.

meatballhat committed:
Retain instance lifecycle actions for some TTL

with a `Completed` boolean to allow for short circuiting,
rather than deleting them on completion.

meatballhat committed:
Remove trailing newline from text response

meatballhat committed:
Back out remaining func map dispatch

and remove the use of the set lookup for lifecycle events, thus saving us from
another potential source of redis leakage.

more tests, too

maybe other stuff.  wow.
